lib/cocoa/src/protocol/TCompactProtocol.m

```diff
@@ -0,0 +1,706 @@
+        }
+        if (pfieldID != NULL)
+        {
+            *pfieldID = 0;
+        }
+        return;
+    }
+    
+    short fieldId = 0;
+    
+    // mask off the 4 MSB of the type header. it could contain a field id delta.
+    short modifier = (byte & 0xf0) >> 4;
+    if (modifier == 0)
+    {
+        // not a delta. look ahead for the zigzag varint field id.
+        fieldId = [self readI16];
+    }
+    else
+    {
+        // has a delta. add the delta to the last Read field id.
+        fieldId = lastFieldId + modifier;
+    }
+    
+    int fieldType = [self getTTypeForCompactType:type];
+    
+    if (pname != NULL)
+    {
+        *pname = @"";
+    }
+    if (pfieldType != NULL)
+    {
+        *pfieldType = fieldType;
+    }
+    if (pfieldID != NULL)
+    {
+        *pfieldID = fieldId;
+    }
+    
+    // if this happens to be a boolean field, the value is encoded in the type
+    if (type == TCType_BOOLEAN_TRUE || type == TCType_BOOLEAN_FALSE)
+    {
+        // save the boolean value in a special instance variable.
+        booleanValue = [[NSNumber numberWithBool:type == TCType_BOOLEAN_TRUE ? YES : NO] retain_stub];
+    }
+    
+    // push the new field onto the field stack so we can keep the deltas going.
+    lastFieldId = fieldId;
+}
+
+-(void)readMapBeginReturningKeyType:(int*)pkeyType valueType:(int*)pvalueType size:(int*)psize
+{
+    uint8_t keyAndValueType = 0;
+    int size = (int)[self readVarint32];
+    if (size != 0)
+    {
+        keyAndValueType = [self readByte];
+    }
+    
+    int keyType = [self getTTypeForCompactType:keyAndValueType >> 4];
+    int valueType = [self getTTypeForCompactType:keyAndValueType & 0xf];
+    
+    if (pkeyType != NULL)
+    {
+        *pkeyType = keyType;
+    }
+    if (pvalueType != NULL)
+    {
+        *pvalueType = valueType;
+    }
+    if (psize != NULL)
+    {
+        *psize = size;
+    }
+}
+
+-(void)readListBeginReturningElementType:(int*)pelementType size:(int*)psize
+{
+    uint8_t size_and_type = [self readByte];
+    int size = (size_and_type >> 4) & 0x0f;
+    if (size == 15)
+    {
+        size = (int)[self readVarint32];
+    }
+    
+    int elementType = [self getTTypeForCompactType:size_and_type & 0x0f];
+    
+    if (pelementType != NULL)
+    {
+        *pelementType = elementType;
+    }
+    if (psize != NULL)
+    {
+        *psize = size;
+    }
+}
+
+-(void)readSetBeginReturningElementType:(int*)pelementType size:(int*)psize
+{
+    [self readListBeginReturningElementType:pelementType size:psize];
+}
+
+-(BOOL)readBool
+{
+    if (booleanValue != nil)
+    {
+        BOOL result = [booleanValue boolValue];
+        [booleanValue release_stub];
+        booleanValue = nil;
+        return result;
+    }
+    else
+    {
+        return [self readByte] == TCType_BOOLEAN_TRUE;
+    }
+}
+
+-(uint8_t)readByte
+{
+    uint8_t buf = 0;
+    [mTransport readAll:&buf offset:0 length:1];
+    return buf;
+}
+
+-(short)readI16
```

Comment:
This should be `int16_t` since it deals in a specific integer size, rather than relying on the fact that for a specific target `short` happens to be 16-bit.